driver-legacy/src/main/com/mongodb/DBCollection.java

```diff
@@ -137,9 +139,10 @@ public class DBCollection {
     private volatile ReadPreference readPreference;
     private volatile WriteConcern writeConcern;
     private volatile ReadConcern readConcern;
+    private final Lock factoryAndCodecLock = new ReentrantLock();
     private DBEncoderFactory encoderFactory;
     private DBDecoderFactory decoderFactory;
-    private DBCollectionObjectFactory objectFactory;
+    private volatile DBCollectionObjectFactory objectFactory;
```

Comment:
An obvious fix here, but it means nothing with regard to whether the rest of the code is correctly synchronized.